### PR TITLE
feat(landing): add glossary

### PR DESCRIPTION
**Pull request by p6l-richard**
## What does this PR do?
Demo 👇
https://33fa1ur95-b7flgfdfq-unkey.vercel.app/glossary
https://33fa1ur95-b7flgfdfq-unkey.vercel.app/glossary/mime-types

CleanShot 2024-10-23 at 17 17 03

Adding a `/glossary` to unkey's landing.

Fixes # (issue)

*If there is not an issue for this, please create one first. This is used to tracking purposes and also helps use understand why this PR exists*

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Test A
- Test B

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [ ] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

## Release Notes

- **New Features**
  - Introduced a comprehensive glossary section with a `GlossaryTermWrapper` for detailed term pages.
  - Added a `GlossaryClient` for an interactive glossary interface, featuring search and category navigation.
  - Implemented responsive components for both desktop (`TermsRolodexDesktop`) and mobile (`TermsStepperMobile`) views for glossary term navigation.
  - Added an FAQ section and a Takeaways component to enhance user engagement with glossary terms.
  - Introduced a FilterableCommand component for searching glossary terms.

- **Documentation**
  - New content file created for MIME types, providing an in-depth guide and best practices.

- **UI Enhancements**
  - Introduced new UI components: `Badge`, `Button`, `Card`, `Input`, `Label`, and `KeyIcon` for improved styling and user interaction.

### CODE REVIEW

File: apps/www/app/glossary/[slug]/_faq.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      <div className="mx-auto md:max-w-[800px]">
+        <Accordion type="single" collapsible className="w-full">
+          {props.items.map((item) => (
+            <AccordionItem value={item.question}>
+              <AccordionTrigger className="justify-between space-x-2">
+                {item.question}
+              </AccordionTrigger>
+              <AccordionContent>{item.answer}</AccordionContent>
+            </AccordionItem>
+          ))}
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/www/app/glossary/[slug]/_faq.tsx:33]: yup, this needs a key

File: apps/www/app/glossary/[slug]/_faq.tsx
Comment (chronark) [inline apps/www/app/glossary/[slug]/_faq.tsx:1]: why the _ prefix on the file name?
is that just to differentiate it from the nextjs named files?

File: apps/www/app/glossary/[slug]/_takeaways.tsx [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+              </li>
+            ) : (
+              <div key={item.key} className="flex justify-between text-sm space-y-2">
+                <span className="font-medium text-white/80">{item.key}</span>
+                <span className="text-white/60">{item.value}</span>
+              </div>
+            ),
+          )}
+        </div>
+      )}
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/www/app/glossary/[slug]/_takeaways.tsx:166]: that seems like an issue

File: apps/www/app/glossary/[slug]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  allGlossaries.map((term) => ({
+    slug: term.slug,
+  }));
+
+export function generateMetadata({
+  params,
+}: {
+  params: { slug: string };
+}): Metadata {
+  const term = allGlossaries.find((term) => term.slug === `${params.slug}`);
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/www/app/glossary/[slug]/page.tsx:32]: yes please

File: apps/www/app/glossary/[slug]/page.tsx [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+      creator: "@unkeydev",
+    },
+    icons: {
+      shortcut: "/images/landing/unkey.png",
+    },
+  };
+}
+
+const GlossaryTermWrapper = async ({ params }: { params: { slug: string } }) => {
+  const term = allGlossaries.find((term) => term.slug === `${params.slug}`) as Glossary;
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/www/app/glossary/[slug]/page.tsx:60]: yes please

File: apps/www/components/glossary/codeblock.tsx
Comment (chronark) [inline apps/www/components/glossary/codeblock.tsx:1]: what's the reason why we don't reuse the existing code block component?

**Review by chronark — COMMENTED** [review]
CleanShot 2024-10-23 at 11 27 49@2x
sidebar still has placeholder names

___

CleanShot 2024-10-23 at 11 28 29@2x

are we keeping the mask icon here? looks a bit weird

File: apps/www/components/glossary/darkTheme.tsx
Comment (chronark) [inline apps/www/components/glossary/darkTheme.tsx:1]: why do we need this?
we already have code blocks with syntax highlighting afaik

File: apps/www/components/glossary/mdx-components.tsx
Comment (chronark) [inline apps/www/components/glossary/mdx-components.tsx:1]: why don't we reuse the blog mdx components?

File: apps/www/components/glossary/terms-navigation-desktop.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+          {visibleTerms.map((term) => (
+            <Link
+              href={`/glossary/${term.slug}`}
+              className={cn(
+                "flex items-center px-2 h-10 rounded-md text-white/60 hover:text-white",
+                {
+                  "text-white": slug === term.slug,
+                },
+              )}
+            >
```
[the remark this answers is not in the document]
Comment (chronark) [reply apps/www/components/glossary/terms-navigation-desktop.tsx:94]: yes please

**Review by chronark — CHANGES_REQUESTED** [review]

File: apps/www/components/glossary/mdx-components.tsx
Comment (p6l-richard) [reply apps/www/components/glossary/mdx-components.tsx:1]: We do actually. These components got copied over when I copied the templates, which have their custom react markdown implementation for the readmes:
https://github.com/p6l-richard/unkey/blob/main/apps/www/app/templates/%5Bslug%5D/page.tsx

removed this: https://github.com/unkeyed/unkey/commit/9144569ecc4238a18fb8317d8e4710d5fa71b9b3

File: apps/www/app/glossary/[slug]/_faq.tsx
Comment (p6l-richard) [reply apps/www/app/glossary/[slug]/_faq.tsx:1]: it keeps adds sorting to the components.
NextJS magic below, components above

I can remove the underscores though to stay consistent with the other namings.

File: apps/www/components/glossary/search.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+export function FilterableCommand(props: {
+  placeholder: string;
+  className?: string;
+  terms: Array<Glossary>;
+}) {
+  const [open, setOpen] = React.useState(false);
+  const router = useRouter();
+  const commandRef = React.useRef<HTMLDivElement>(null);
+
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/search.tsx:24]: nah

File: apps/www/components/glossary/search.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                value={item.slug}
+                className="px-3 py-2 cursor-pointer flex items-center w-full text-sm text-white/60 hover:text-white"
+                onSelect={() => router.push(`/glossary/${item.slug}`)}
+              >
+                {item.title}
+              </CommandPrimitive.Item>
+            ))}
+          </CommandGroup>
+        </CommandList>
+      )}
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/search.tsx:57]: not needed

File: apps/www/components/glossary/terms-rolodex-desktop.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    const halfVisible = Math.floor(totalVisible / 2);
+
+    const start = currentIndex - halfVisible;
+    const end = currentIndex + halfVisible + 1;
+
+    const wrappedTerms = [...terms, ...terms, ...terms];
+    const centerOffset = terms.length;
+
+    return wrappedTerms.slice(centerOffset + start, centerOffset + end);
+  };
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/terms-rolodex-desktop.tsx:37]: react compiler will eventually handle that stuff

File: apps/www/content/glossary/mime-types.mdx [hunk 9]
```diff
@@ -0,0 +1,100 @@
+---
+title: "MIME Types Explained"
+description: "Learn about MIME types in API development. Understand their role in defining file formats like images and PDFs. Explore our glossary for more insights."
+term: "MIME Types"
+h1: "What are MIME Types? Format IDs Explained"
+intro: "MIME types are essential identifiers in digital communication, specifying the nature and format of data. They play a crucial role in web development and API interactions, ensuring that information is correctly interpreted and displayed across various platforms and applications."
+reviewer: andreas
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/content/glossary/mime-types.mdx:7]: nah

File: apps/www/components/glossary/search.tsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        // The `onBlur` event checks if the new focus target (relatedTarget) is within the Command component:
+        // - If it's not (i.e., clicking outside), it closes the list.
+        // - If it is (i.e., selecting an item), it keeps the list open, allowing the `onSelect` to handle the navigation.
+        onBlur={(event: React.FocusEvent<HTMLInputElement>) => {
+          const relatedTarget = event.relatedTarget as Node | null;
+          if (!commandRef.current?.contains(relatedTarget)) {
+            setOpen(false);
+          }
+        }}
+      />
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/search.tsx:39]: fair suggestion, but not sure if this isn't added under the hood by cmd+k?

File: apps/www/components/glossary/terms-stepper-mobile.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            href={`/glossary/${nextTerm.slug}`}
+            className="flex w-full flex-row items-center gap-2 rounded-lg p-4 text-sm transition-colors text-white/60 hover:text-white"
+          >
+            {nextTerm.title}
+            <ChevronRightIcon className="size-4 shrink-0" />
+          </Link>
+        </div>
+      </div>
+    </div>
+  );
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/terms-stepper-mobile.tsx:48]: fair

File: apps/www/components/glossary/terms-stepper-mobile.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  terms,
+}: { className?: string; terms: Array<Pick<Glossary, "slug" | "title">> }) {
+  const params = useParams();
+  const slug = params.slug as string;
+  const sortedTerms = terms.sort((a, b) => a.title.localeCompare(b.title));
+  const slugIndex = sortedTerms.findIndex((term) => term.slug === slug);
+  const startIndex = slugIndex !== -1 ? slugIndex : 0;
+  const currentTerm = sortedTerms[startIndex];
+  const previousTerm = sortedTerms[(startIndex - 1 + sortedTerms.length) % sortedTerms.length];
+  const nextTerm = sortedTerms[(startIndex + 1) % sortedTerms.length];
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/glossary/terms-stepper-mobile.tsx:20]: not overcomplicate

File: apps/www/components/glossary/codeblock.tsx
Comment (p6l-richard) [reply apps/www/components/glossary/codeblock.tsx:1]: We do actually. These components got copied over when I copied the templates, which have their custom react markdown implementation for the readmes:
https://github.com/p6l-richard/unkey/blob/main/apps/www/app/templates/%5Bslug%5D/page.tsx

removed this: https://github.com/unkeyed/unkey/commit/9144569ecc4238a18fb8317d8e4710d5fa71b9b3

File: apps/www/app/glossary/[slug]/_faq.tsx
Comment (p6l-richard) [reply apps/www/app/glossary/[slug]/_faq.tsx:1]: removed prefix: https://github.com/unkeyed/unkey/pull/2085/commits/312075bafd61a7569780991494a36460c9ec7e3c

File: apps/www/components/ui/button.tsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
+  ({ className, variant, size, asChild = false, ...props }, ref) => {
+    const Comp = asChild ? Slot : "button";
+    return (
+      <Comp className={cn(buttonVariants({ variant, size, className }))} ref={ref} {...props} />
+    );
+  },
+);
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/components/ui/button.tsx:46]: shadcn

File: apps/www/app/glossary/[slug]/takeaways.tsx [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+  ),
+  didYouKnow: z.string(),
+});
+
+export default function Takeaways(props: z.infer<typeof takeawaysSchema>) {
+  return (
+    <Card className="w-full bg-white/5 shadow-[0_0_10px_rgba(255,255,255,0.1)] rounded-xl overflow-hidden relative border-white/20">
+      <div className="absolute left-0 top-0 bottom-0 w-1 bg-white/20" />
+      <CardHeader className="border-white/20">
+        <CardTitle className="text-2xl font-bold text-white">MIME Types: Key Takeaways</CardTitle>
```
[the remark this answers is not in the document]
Comment (p6l-richard) [reply apps/www/app/glossary/[slug]/takeaways.tsx]: good

**Review by chronark — APPROVED** [review]

**Pull request merged**